polus-stack-z-slice-plugin/src/main.py

```diff
@@ -1,16 +1,95 @@
-import argparse, time, logging, subprocess, multiprocessing
+import argparse, logging, math, filepattern, time, queue
+from bfio import BioReader, BioWriter
 from pathlib import Path
-from utils import _parse_files_p,_parse_files_xy,_parse_fpattern
+from preadator import ProcessManager
+
+logging.basicConfig(format='%(asctime)s - %(name)-8s - %(levelname)-8s - %(message)s',
+                    datefmt='%d-%b-%y %H:%M:%S')
+
+# length/width of the chunk each _merge_layers thread processes at once
+chunk_size = 8192
+
+# Units for conversion
+UNITS = {'m':  10**9,
+         'cm': 10**7,
+         'mm': 10**6,
+         'µm': 10**3,
+         'nm': 1,
+         'Å':  10**-1}
+
+def _merge_layers(input_files,output_path):
+    
+    with ProcessManager.process(output_path.name):
+
+        # Get the number of layers to stack
+        z_size = 0
+        for f in files:
+            with BioReader(f['file']) as br:
+                z_size += br.z
+                
+        # Get some basic info about the files to stack
+        with BioReader(input_files[0]['file']) as br:
+
+            # Get the physical z-distance if avaiable, set to physical x if not
+            ps_z = br.ps_z
+            
+            # If the z-distances are undefined, average the x and y together
+            if None in ps_z:
+                # Get the size and units for x and y
+                x_size,x_units = br.ps_x
+                y_size,y_units = br.ps_y
+                
+                # Convert x and y values to the same units and average
+                z_size = (x_size*UNITS[x_units] + y_size*UNITS[y_units])/2
+                
+                # Set z units to the smaller of the units between x and y
+                z_units = x_units if UNITS[x_units] < UNITS[y_units] else y_units
+                
+                # Convert z to the proper unit scale
+                z_size /= UNITS[z_units]
+                ps_z = (z_size,z_units)
+                ProcessManager.log('Could not find physical z-size. Using the average of x & y {}.'.format(ps_z))
+
+            # Hold a reference to the metadata once the file gets closed
+            metadata = br.metadata
+
+        # Create the output file within a context manager
+        with BioWriter(output_path,metadata=metadata,max_workers=ProcessManager._active_threads) as bw:
+
+            # Adjust the dimensions before writing
+            bw.z = z_size
+            bw.ps_z = ps_z
+
+            # ZIndex tracking for the output file
+            zi = 0
+
+            # Start stacking
+            for file in input_files:
+
+                with BioReader(file['file'],max_workers=ProcessManager._active_threads) as br:
+
+                    for z in range(br.z):
+
+                        for xs in range(0,br.x,chunk_size):
```

Comment:
Consider adding a comment to explain the stacking process in more detail or more descriptive variable naming of xs, yx, xe, xs